Adding IsProximityHovered property of type TimedFlag to detect when a button starts being hovered or on interactor proximity and when it stops being hovered or on proximity of any interactor.

ms-RistoRK opened the pull request:
This PR is to adds IsProximityHovered property of type TimedFlag to detect when a button starts being hovered or on interactor proximity and when it stops being hovered or on proximity of any interactor.
Such property will be used in the near future to make the FrontPlate's RawImage to be dynamically enabled/disabled based on Proximity + Hovering Enter/Exit events.

* The motivation of such new behavior is the research done to find bottlenecks on MRTK3 buttons and increase the amount of concurrent buttons in a scene before a significant framerate drop (60 FPS).
* The impact of this change is that MRTK3 scenes will now be able to support ~84% of buttons as in MRTK2 scenes (current is ~46%), effectively doubling the amount of buttons before a significant framerate drop (also documented).
* The philosophy behind this code change is:
  * We retain the investment previously done on the improved visual appeal of MRTK3 buttons, nothing is erased.
* New UnityTests for the above ^ are also included in this PR.

keveleigh commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -623,6 +672,33 @@ private float IntersectQuad(Vector3 start, Vector3 end, float depth)
+        /// <param name="enable">True to enable, false to disable</param>
+        protected void SetEnableFrontPlateRawImage(bool enable)
+        {
+            GameObject frontPlateGameObject = transform.Find(FRONT_PLATE_NAME).gameObject;
+            UnityEngine.UI.RawImage frontPlateRawImage = frontPlateGameObject?.GetComponent<UnityEngine.UI.RawImage>();
```
Can we cache these image references on startup instead of calling `Find` and `GetComponent` on every call?

ms-RistoRK replied:
I initially thought on something similar but then I realized that such change would also require to implement a way to update the cached image reference if someone develops an app in which the Front Plate Raw Images are updated in runtime.  Then I thought that a Find + GetComponent call would increase the computational complexity
with a constant O(k) which is acceptable if it doesn't happen often (it doesn't happen on a per-frame basis).  So, considering it doesn't impact performance significantly and aiming to keep the code change simple I thought on implementing it as I initially suggested.  May I know your thoughts on my reasoning?  Also, if your company has a specific need to have it cached that we are not aware of then please let us know.

anonymous2585 replied:
I really don't like referencing objects by name (= the use of transform.Find() or GameObject.Find()), even when cached.
What if a user when to rename the frontplate gameobject?

marlenaklein-msft replied:
An alternative may be to use a serialized field (like how InteractablePulse stores the front plate).

ms-RistoRK replied:
Thank you all for your feedback, I've tried to capture it in the latest commit.

keveleigh commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -623,6 +672,33 @@ private float IntersectQuad(Vector3 start, Vector3 end, float depth)
+        /// <param name="enable">True to enable, false to disable</param>
+        protected void SetEnabledCanvasElementRoundedRectIfAny(bool enable)
+        {
+            Microsoft.MixedReality.GraphicsTools.CanvasElementRoundedRect canvasElementRoundedRect = gameObject.FindAncestorComponent<Microsoft.MixedReality.GraphicsTools.CanvasElementRoundedRect>();
```
This feels like it could be cached as well.

whebertML replied:
It isn't odd that a child element would be directly modifying the enabled state of some parent component higher up in the hierarchy based on the child element's hover events?  At least built-in like this.  If that is needed in a particular use case, couldn't a dev instead hookup such things with events in the Inspector?

ms-RistoRK replied:
Cached in latest commit, thank you for the feedback.

keveleigh commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -148,6 +150,20 @@ public SpaceMode DistanceSpaceMode
+        /// </summary>
+        [SerializeField]
+        [Tooltip("Enables (true) or disables (false) the Canvas Rounded Rect on hover only.")]
+        public bool EnableOnHoverOnlyCanvasRoundedRect = false;
+
+        /// <summary>
+        /// Flag that enables (true) or disables (false) the Front plate Raw Image on hover only.
+        /// </summary>
+        [SerializeField]
+        [Tooltip("Enables (true) or disables (false) the Front plate Raw Image on hover only.")]
+        public bool EnableOnHoverOnlyFrontPlateRawImage = false;
```
Although I don't think we yet have explicit coding guidelines for MRTK3, we've thus far enforced some similar rules to MRTK2, including "Fields should always be private, with public or protected property accessors."

https://learn.microsoft.com/en-us/windows/mixed-reality/mrtk-unity/mrtk2/contributing/coding-guidelines?view=mrtkunity-2022-05#access-modifiers

Please update these fields to `private` and adding public fields pointing to them, if needed.

keveleigh replied:
You could also convert them into auto-properties and assign the attribute directly to the backing field with `[field: SerializeField]` :

https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/blob/1a61719eb59fa5042db916dbb031f8c3ff9eefe5/org.mixedrealitytoolkit.core/Interactables/StatefulInteractable.cs#L48-L50

ms-RistoRK replied:
Thx for the feedback, I wasn't aware of those guidelines, I'll update accordingly.

ms-RistoRK replied:
Done in latest commit.

keveleigh commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -279,6 +295,21 @@ protected override void Reset()
+        /// <summary>
+        /// A Unity event function that is called when the script is loaded.
+        /// </summary>
+        private void Start()
```
To match the other Unity functions in this file, as well as to prevent derived classes from having their calls blocked, it would be good to make this `protected` and probably `virtual`.

For more info: https://github.com/microsoft/Microsoft.Unity.Analyzers/blob/main/doc/UNT0021.md

A downside to this change is it'd _technically_ be a breaking change if anybody has created their own derived classes with a defined `Start()` method. It might be worth discussing how we want to approach this.

ms-RistoRK replied:
You are right, thanks for the feedback.  Updated in latest commit.

anonymous2585 commented:
Is it possible to make the docs and videos public?
Right now, it's required to log with a Microsoft account, and I don't have one.

whebertML commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -148,6 +150,12 @@ public SpaceMode DistanceSpaceMode
         [Tooltip("Speed for extending the moving button visuals when selected by a non-touch source.")]
         private float extendSpeed = 0.5f;
 
+        [field: SerializeField, Tooltip("Enables (true) or disables (false) the Canvas Rounded Rect on hover only.")]
+        public bool EnableOnHoverOnlyCanvasRoundedRect { get; set; } = false;
+
+        [field: SerializeField, Tooltip("Enables (true) or disables (false) the Front plate Raw Image on hover only.")]
+        public bool EnableOnHoverOnlyFrontPlateRawImage { get; set; } = false;
+
```
Wouldn't adding such visual concepts directly to PressableButton break the philosophy of this component, in that it (from \<remarks\> above) /// This script does not make any assumptions about the visuals associated with this button.

AMollis replied:
I tend to agree with this.  Can enabling and disabling be controlled event a event handler within the prefab itself?

ms-RistoRK replied:
> I tend to agree with this. Can enabling and disabling be controlled event a event handler within the prefab itself?

I don't understand what you mean by this.  Can you provide an example of what you mean?

ms-RistoRK replied:
Thank you all for your feedback.  I have switched from OnHover* events to proximity interactor similar to how proximity hover lights work.  Would you agree with this new approach?

whebertML reviewed (commented):
Can't see the docs or videos showcasing the reason for this change, but it would appear to possibly break some design concepts for PressableButton?

AMollis reviewed (changes requested):
I think we need to use another state than hovering.   With hovering you miss the scenario where an interactor is in proximity of many buttons.  In this case we should still show the front plate so that the visual indication is shown.

For example, this image shows the scenario I'm talking about (screen shot from video without your change):

<img width="125" alt="image" src="https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/36461279/9f6fa0ad-7ede-41ff-96da-cfdf263cbcad">

ms-RistoRK commented:
> I think we need to use another state than hovering. With hovering you miss the scenario where an interactor is in proximity of many buttons. In this case we should still show the front plate so that the visual indication is shown.
> 
> For example, this image shows the scenario I'm talking about (screen shot from video without your change):
> 
> <img alt="image" width="125" src="https://private-user-images.githubusercontent.com/36461279/296136556-9f6fa0ad-7ede-41ff-96da-cfdf263cbcad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDUwODA0NzcsIm5iZiI6MTcwNTA4MDE3NywicGF0aCI6Ii8zNjQ2MTI3OS8yOTYxMzY1NTYtOWY2ZmEwYWQtN2VkZS00MWZmLTk2ZGEtY2ZkZjI2M2NiY2FkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTEyVDE3MjI1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY5ZDVlMzQ3MzdiZGE5MzBkN2QzZmUyYzc5M2QwZWU4NmI3NzRkMTI4ODIxM2Y0OTgyOTY4MGU1MzVhM2RkZjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.x8Gm76aWGFnB0KBu6qwBBFkrEsluz1N_a9EgYFiMuks">

Ok, so, which other state do you want me to use?

ms-RistoRK commented:
> Is it possible to make the docs and videos public? Right now, it's required to log with a Microsoft account, and I don't have one.

I am afraid not.  I removed them to avoid confusions.

AMollis commented:
> > I think we need to use another state than hovering. With hovering you miss the scenario where an interactor is in proximity of many buttons. In this case we should still show the front plate so that the visual indication is shown.
> > For example, this image shows the scenario I'm talking about (screen shot from video without your change):
> > <img alt="image" width="125" src="https://private-user-images.githubusercontent.com/36461279/296136556-9f6fa0ad-7ede-41ff-96da-cfdf263cbcad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDUwODA0NzcsIm5iZiI6MTcwNTA4MDE3NywicGF0aCI6Ii8zNjQ2MTI3OS8yOTYxMzY1NTYtOWY2ZmEwYWQtN2VkZS00MWZmLTk2ZGEtY2ZkZjI2M2NiY2FkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTEyVDE3MjI1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY5ZDVlMzQ3MzdiZGE5MzBkN2QzZmUyYzc5M2QwZWU4NmI3NzRkMTI4ODIxM2Y0OTgyOTY4MGU1MzVhM2RkZjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.x8Gm76aWGFnB0KBu6qwBBFkrEsluz1N_a9EgYFiMuks">
> 
> Ok, so, which other state do you want me to use?

I think this is something we may need to create.  How are proximity hover lights working today?  I would look into what is controlling that, can base front plate enablement based on whatever is rendering hover lights

ms-RistoRK commented:
@keveleigh, @whebertML, @anonymous2585: Thank you all for your feedback and for taking the time to look into this PR.  I cannot share the internal docs regarding the benchmarking we did and its findings because they are labeled as "Confidential".  However, I can tell you in general terms what we did and what motivated this PR:
1 - We developed a benchmarking tool similar to MRTK2's PerformanceMeasurement tool and we used it to profile UI elements and try to find performance bottlenecks.
2 - Our experiments showed that if we disabled the Front plate Raw Image and Canvas Rounded Rect then we would double the amount of buttons that an MRTK3 scene could support in an HL2 device before falling below a 60 fps rate.
3 - We then though on a way to allow developers to have an easy way to disable such components without having them to do it manually.
Based on the above, we created this PR so that developers don't lose any of the existing behaviors nor visual appeal yet they have an easy way to activate the new behavior and be able to have more buttons in their UIs before a significant framerate occurs.
I hope this gives you more information on the motivation of this PR.  Please feel free to reach out if you have any further comments or questions.

ms-RistoRK commented:
Update:  I am making this PR back to draft while I figure out why some tests are failing.

ms-RistoRK commented:
Re-opening this PR as all tests are now passing.

AMollis commented on org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs:
```diff
@@ -20,10 +30,38 @@ public class NearInteractionModeDetector : ProximityDetector
+                    Collider previouslyDetectedCollider = previouslyDetectedColliders[i];
+                    if (!DetectedColliders.Contains(previouslyDetectedCollider) && previouslyDetectedCollider != null)
+                    {
+                        previouslyDetectedCollider.GetComponent<IPressableButtonForNearInteractionModeDetector>()?.UpdateFrontPlateAndRoundedRectIfDynamic(false);
```
The `?` operator will fail if the previous collider was destroyed.  Unity destroys the underlying native objects when an object is destroyed, but the c# object will still be non-null.   To correctly handle this you need to use the == or != operators.  Unity has overrides for these, but not the ? operator

https://docs.unity3d.com/ScriptReference/Object-operator_eq.html

AMollis replied:
same comment for line 59

ms-RistoRK replied:
Thank you for the explanation.  Updated accordingly in latest commit.  Resolving this conversation.

AMollis commented on org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs:
```diff
@@ -20,10 +30,38 @@ public class NearInteractionModeDetector : ProximityDetector
+                }
+                foreach (Collider collider in DetectedColliders)
+                {
+                    if (!previouslyDetectedColliders.Contains(collider))
```
using a hash set for `previouslyDetectedColliders` will be more performant.   HashSet<Collider>

ms-RistoRK replied:
Thank you for the feedback.  As we discussed offline, the collection gets modified by other instances, for example if I switch from a List to a Hashset as in:
![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/9d2812e1-ad17-4b4c-8ece-352e023801c4)
which I would traverse with:
![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/a9de0730-da8e-4ea1-950e-aeb902163215)
then I get several exceptions as this:
![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/da7a6567-4ca5-4ce7-a253-22488a4df36e)
I could make a copy of the Hashset to an array and traverse it with a for cycle but that would make it less performant and increase the memory consumption.
Because of the above, as we discussed offline, I will keep the use of a List.  I do appreciate your feedback and the time you invested in reviewing this PR though.  Resolving this conversation.

AMollis commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -279,6 +317,28 @@ protected override void Reset()
+        /// </summary>
+        protected virtual void Start()
+        {
+            canvasElementRoundedRect = gameObject.FindAncestorComponent<CanvasElementRoundedRect>();
```
Why can't we just disable the enable FrontPlate game objects?  Why do we need to target the components?

AMollis replied:
If we need to target the components, then we should either pass those in as serialized fields.

ms-RistoRK replied:
The GameObject has other components that contributed to the button's visual appeal that I didn't want to lose hence I only targeted the CanvasElementRoundedRect, also the experiments to find rendering bottlenecks showed the CanvasElementRoundedRect was a main bottleneck.  Resolving this conversation as the new path is to create an array of component where the user can decide which will be dynamic based on proximity, as per updated User Story.

AMollis commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -148,6 +152,28 @@ public SpaceMode DistanceSpaceMode
         private float extendSpeed = 0.5f;
 
+        [field: SerializeField, Tooltip("Enables (true) or disables (false) the Canvas Rounded Rect on proximity only.")]
+        public bool EnableOnProximityOnlyCanvasRoundedRect { get; set; } = false;
```
If we are passing in components here (i.e. FrontPlate)  I don't think we need to have these EnableOnProximityOnlyCanvasRoundedRect or EnableOnProximityOnlyFrontPlateRawImage fields.

If someone wanted to opt of the front plate disablement, they can leave this FrontPlate field empty

ms-RistoRK replied:
Makes sense, I'll update accordingly, thx for the feedback.  Actually, since the new direction is to use an array of components in which the user decides which components to make dynamic based on proximity (as per udpated User Story) I will remove these fields.

ms-RistoRK replied:
In the last commit I removed these fields as the new behavior is to enable/disable by proximity any Component in such array as per new requirements.  Resolving this conversation.

AMollis commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -148,6 +152,28 @@ public SpaceMode DistanceSpaceMode
+        /// <summary>
+        /// ActionButton's Front plate.
+        /// </summary>
+        public GameObject FrontPlate
```
I think we should rename this to something less specific.  Perhaps `ProximityObjects`?  And perhaps make it an array to allow for multiple pieces to be hidden?

ms-RistoRK replied:
The original requirement was to only make dynamic Rounded Rect and Front Plate's Raw Image dynamic, however I agree it would be more valuable for the users to be able to make dynamic any component(s) they wish.  I'll update this PR accordingly and update the scope of the User Story in the board.

ms-RistoRK replied:
In the last commit I removed these fields as the new behavior is to enable/disable by proximity any Component in such array as per new requirements.  Resolving this conversation.

AMollis commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -148,6 +152,28 @@ public SpaceMode DistanceSpaceMode
+        /// <summary>
+        /// ActionButton's Front plate.
+        /// </summary>
+        public GameObject FrontPlate
```
I thought we were going down the path of using events to show or hide elements?    Why have you decided to go down this path?

I sort of like using events, so the button logic doesn't need to be aware of the visuals of button.   See `remarks`....a core principle of this component is to make no assumptions about visuals.  If we start making a strong connection to visuals in this class, we risk breaking the contract we had with developers.

ms-RistoRK replied:
In the last commit I switched to the new behavior is to enable/disable by proximity any Component in an array populated in Editor as per new requirements.  I think the need to hook to event handlers won't be needed anymore.  Resolving this conversation but please feel free to reopen if considered necessary.

ms-RistoRK commented on org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs:
```diff
@@ -7,6 +7,16 @@
+    /// button if they are flagged as dynamic (based on proximity), it is needed to prevent
+    /// a circular reference between MRTK Input and MRTK UX Core Scripts packages.
+    /// </summary>
+    public interface IPressableButtonForNearInteractionModeDetector
```
Done and done, implemented in latest commit.  Resolving this conversation.

ms-RistoRK commented on org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs:
```diff
@@ -20,10 +30,38 @@ public class NearInteractionModeDetector : ProximityDetector
         {
-            return base.IsModeDetected() || IsNearInteractorSelecting();
+            bool result = base.IsModeDetected() || IsNearInteractorSelecting();
+            if (result)
```
I didn't test it that way, thank you for pointing it out.  I will test it next and share my findings here.

ms-RistoRK replied:
I finished testing with multiple fingers and you are correct, wrong states can happen.  I'll update the code to deal with those cases.  Thanks for catching that and letting me know.

ms-RistoRK replied:
The last commit adds the functionality to support multiple interactors when dynamically enabling/disabling buttons components.  The new behavior is shown next:
![HL2Demo20240125_001](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/f2ca31c6-8b5a-49cc-b9db-05fd85f10d36)
I also added a runtime test that verifies the new behavior.  Essentially, the new test does this:
![TestRecording](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/7e5e3a8f-fda1-45b7-bd6f-bf8e719107a9)
I believe the commit addresses this issue so resolving this conversation, feel free to re-open if needed.

ms-RistoRK commented on org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs:
```diff
@@ -7,6 +7,16 @@
+    /// </summary>
+    public interface IPressableButtonForNearInteractionModeDetector
+    {
+        public void UpdateFrontPlateAndRoundedRectIfDynamic(bool enable);
```
Thank you for the suggestion but it won’t work as how NearInteractionModeDetector works right now because it only keeps track of the last interactor triggering the near interaction.  For example, I created this test:
![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/37ef8075-6b6d-4e0e-85ff-dc0c9c4a34f7)
Which does this:
1. Creates a dialog with a button
![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/19e978a4-17e8-4937-8893-b12b62534493)
2. Creates a right hand and touches the button
![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/4a7a8e77-4090-45c6-8f9f-d73f1baf66db)
3. Creates a left hand and touches the button
![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/dba709af-3a14-415b-a73d-3236959a84cf)
At all steps the NearInteractors count remain the same, the only difference is that the right hand interactor reference gets replaced with the left hand interactor reference.
I am currently investigating if there is already a way to query or if there is a data structure that holds all active interactors but I am new to this part of the code (if you know of any please let me know).  Otherwise, I will have to add some way to keep track of that but I was hoping to keep this PR simple.

ms-RistoRK replied:
The last commit adds the functionality to support multiple interactors when dynamically enabling/disabling buttons components. I also added a runtime test that verifies the new behavior.
I believe the commit addresses this issue so resolving this conversation, feel free to re-open if needed.

ms-RistoRK commented:
Just a comment to test if my animated gifs are visible:
First:
![HL2Demo20240125_001](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/7f79dc6d-0181-449f-a27d-8599660b59b9)
Second:
![TestRecording](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/ca70a252-123f-430c-90ac-093b8b156a55)

ms-RistoRK commented:
In the last [commit](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/611/commits/e513dca4d547016a5a29c0f0ec2b0e2020af560e) I have:
* Switched to an array of Components to make dynamically enabled/disabled based on proximity.
* Added a new scene to demonstrate the new functionality.  The new scene, CanvasExampleWithDynamicComponents.unity, is a clone of CanvasExample.unity scene but each Action Button, Hero Button, and List Item has dynamic canvas' CanvasElementRoundedRects and dynamic FrontPlate's RawImage Components as those were the target Components to make dynamic in the original PR.
* Updated UnityTests accordingly.

AMollis commented on org.mixedrealitytoolkit.input/InteractionModes/INearInteractionModeTarget.cs:
```diff
@@ -0,0 +1,41 @@
+    /// <summary>
+    /// This interface is used to update the enable state of the Components that are in the ProximityEnabledComponents
+    /// array (set in Editor) and to keep track of which Collider + XRBaseInteractor duples are triggering proximity.
+    /// The interface is needed to prevent a circular reference between MRTK Input and MRTK UX Core Scripts packages.
```
Thank you for the nice comments.

Please put this second sentence in a <remarks> section, and add <see/> tags on types.  To improve readability of the learn.microsoft.com docs

```cs
    /// <summary>
    /// This interface is used to update the enable state of the components that are in the <see cref="ProximityEnabledComponents"/>
    /// array (set in Editor) and to keep track of which <see cref="Collider"/> and <see cref="XRBaseInteractor"/> duples are triggering proximity.
    /// </summary>
    /// <remarks>
    /// The interface is needed to prevent a circular reference between MRTK Input and MRTK UX Core Scripts packages.
    /// </remarks>
```

ms-RistoRK replied:
Updated in last commit.  Resolving.

AMollis commented on org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs:
```diff
@@ -20,10 +20,57 @@ public class NearInteractionModeDetector : ProximityDetector
+                            {
+                                previouslyDetectedCollider.GetComponent<INearInteractionModeTarget>().UnregisterActiveColliderWithInteractor(previouslyDetectedCollider, xrBaseInteractor);
+                            }
+                            if (previouslyDetectedCollider.GetComponent<INearInteractionModeTarget>().ActiveColliderWithInteractorCount == 0)
```
I would like some other opinions here (maybe @keveleigh, @david-c-kline , @whebertML , @ghazen-ml , @shaynie )....

But I think we can safely push the `SetEnableDynamicComponents` out of the `INearInteractionModeTarget` interface, and force the interface implementation to call `SetEnableDynamicComponents` itself.   That is, I'm in favor of a simpler interface.

ms-RistoRK replied:
I cannot do that (I tried though).  If I remove them from the interface then the NeartInteractionModeDetector is unable to call them directly from the PressableButton instance because that would require to NearInteractionModeDetector to know PressableButton type and that would generate a circular reference between packages.

ms-RistoRK replied:
As per offline discussion, I misunderstood the original suggestion as just removing the methods from the interface.  I have removed the methods from the interface and moved the logic to update the state to PressableButton class.  Resolving this.

AMollis commented on org.mixedrealitytoolkit.uxcomponents.noncanvas/Tests/Runtime/PressableButtonTests.cs:
```diff
@@ -850,6 +853,215 @@ public IEnumerator ChangeSpeechSettingsDisabled([ValueSource(nameof(PressableBut
+        /// This tests verifies the that the ProximityEnabledComponents array is empty by default.
+        /// </summary>
+        [UnityTest]
+        public IEnumerator TestPressableButtonProximityEnabledComponentsArrayIsEmptyByDefault([ValueSource(nameof(PressableButtonsTestPrefabPaths))] string prefabFilename)
```
Thank you for the Unity Tests!

ms-RistoRK replied:
No problemo :)

AMollis commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -142,12 +145,18 @@ public SpaceMode DistanceSpaceMode
+        /// Array with Components (set in Editor) that will be enabled/disabled based on proximity.
+        /// </summary>
+        [field: SerializeField, Tooltip("Array with Components (set in Editor) that will be enabled/disabled based on proximity.")]
+        public Component[] ProximityEnabledComponents { get; set; } = null;
```
Maintainers (@keveleigh , @david-c-kline , @whebertML , @ghazen-ml , @shaynie ), I'd like your opinion on this.

I more in favor of adding `InProximityStarted`  and `InProximityStopped` events, and adding event handlers to hide/show button components.   Such events would fall more inline with the existing `StatefulInteractable` events.  

Also, different prefabs may want to do something other than enable/disable components on proximity.  Such events would make that possible.

ms-RistoRK replied:
We could have both :) :   Add event handlers for InProximityStarted and InProximityStopped events so the customers can create custom behaviors and use the Component[] ProximityEnabledComponents array for the default enable/disable on proximity behavior.

whebertML replied:
@ms-RistoRK @AMollis 
Clarification, are we intending to also turn off the whole parent plate background (the blue background for all elements as seen in gif below).  I initially thought this was only to enable/disable the proximity highlight and rounded outline of each element that reacts to proximity - not the parent rounded plate background as well.
![MRTK3_DynamicCanvas1](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/58860572/74b79f62-7958-4aa2-92c6-c7ebeb043502)

ms-RistoRK replied:
> @ms-RistoRK @AMollis Clarification, are we intending to also turn off the whole parent panel background (the blue background for all elements as seen in gif below). I initially thought this was only to enable/disable the proximity highlight and rounded outline of each element that reacts to proximity - not the parent rounded panel background as well. ![MRTK3_DynamicCanvas1](https://private-user-images.githubusercontent.com/58860572/301676148-74b79f62-7958-4aa2-92c6-c7ebeb043502.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY4MzQ3MDcsIm5iZiI6MTcwNjgzNDQwNywicGF0aCI6Ii81ODg2MDU3Mi8zMDE2NzYxNDgtNzRiNzlmNjItNzk1OC00YWEyLTkyYzYtYzdlYmViMDQzNTAyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjAyVDAwNDAwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJkMTAwODU2ZTY3MzViOTNkMzZmNGU2YmUxM2VjZDVmMzllYmVjMTBmZGZjYzkwZTBhM2U5ODEwZmU5NTYxOWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jp4IVFd-mQaKDw-De2YLH49NE8TQA4Dc9zPwK6i2Jhs) [ ![MRTK3_DynamicCanvas1](https://private-user-images.githubusercontent.com/58860572/301676148-74b79f62-7958-4aa2-92c6-c7ebeb043502.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY4MzQ3MDcsIm5iZiI6MTcwNjgzNDQwNywicGF0aCI6Ii81ODg2MDU3Mi8zMDE2NzYxNDgtNzRiNzlmNjItNzk1OC00YWEyLTkyYzYtYzdlYmViMDQzNTAyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjAyVDAwNDAwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJkMTAwODU2ZTY3MzViOTNkMzZmNGU2YmUxM2VjZDVmMzllYmVjMTBmZGZjYzkwZTBhM2U5ODEwZmU5NTYxOWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jp4IVFd-mQaKDw-De2YLH49NE8TQA4Dc9zPwK6i2Jhs) ](https://private-user-images.githubusercontent.com/58860572/301676148-74b79f62-7958-4aa2-92c6-c7ebeb043502.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY4MzQ3MDcsIm5iZiI6MTcwNjgzNDQwNywicGF0aCI6Ii81ODg2MDU3Mi8zMDE2NzYxNDgtNzRiNzlmNjItNzk1OC00YWEyLTkyYzYtYzdlYmViMDQzNTAyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjAyVDAwNDAwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJkMTAwODU2ZTY3MzViOTNkMzZmNGU2YmUxM2VjZDVmMzllYmVjMTBmZGZjYzkwZTBhM2U5ODEwZmU5NTYxOWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jp4IVFd-mQaKDw-De2YLH49NE8TQA4Dc9zPwK6i2Jhs) [ ](https://private-user-images.githubusercontent.com/58860572/301676148-74b79f62-7958-4aa2-92c6-c7ebeb043502.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY4MzQ3MDcsIm5iZiI6MTcwNjgzNDQwNywicGF0aCI6Ii81ODg2MDU3Mi8zMDE2NzYxNDgtNzRiNzlmNjItNzk1OC00YWEyLTkyYzYtYzdlYmViMDQzNTAyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjAyVDAwNDAwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJkMTAwODU2ZTY3MzViOTNkMzZmNGU2YmUxM2VjZDVmMzllYmVjMTBmZGZjYzkwZTBhM2U5ODEwZmU5NTYxOWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jp4IVFd-mQaKDw-De2YLH49NE8TQA4Dc9zPwK6i2Jhs)

@whebertML Thank you for looking into this.  The requirements have changed from the original PR.  Making the canvas' CanvasRoundedRect component dynamic is not a requirement anymore.  Actually, the new approach is that the user selects which component(s) to make dynamic on proximity.
Note 1:  No button prefabs have any components set as dynamic btw.  The original behavior is kept as before this PR.
Note 2:  A new requirement to make components dynamic based on OnHover* and OnFocus* events just came in today which I will implement in the following days.
Note 3:  I updated CanvasExampleWithDynamicComponents.unity scene so that only the button's FrontPlate RawImage component is dynamic, the canvas' CanvasRoundedRect is no longer set as dynamic (also as per new requirement).

whebertML replied:
@ms-RistoRK   Can you possibly share the new requirements?  Sounds like it will be using an event-based system like what @AMollis suggested above?  My personal opinion is that that would be more desirable than making PressableButton implement something like INearInteractionModeTarget.

It seems more ideal to provide such generic events for an Interactable that the dev could use to do whatever, including making certain visual components enabled/disabled like in this case - and so something like PressableButton need not even know about it.

ms-RistoRK replied:
> @ms-RistoRK Can you possibly share the new requirements? Sounds like it will be using an event-based system like what @AMollis suggested above? My personal opinion is that that would be more desirable than making PressableButton implement something like INearInteractionModeTarget.
> 
> It seems more ideal to provide such generic events for an Interactable that the dev could use to do whatever, including making certain visual components enabled/disabled like in this case - and so something like PressableButton need not even know about it.

I'll make the change to an event-based approach.  As per the requirements, the original requirements were:
![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/088d1b45-6e14-4f3c-8331-15a9d16055eb)
This have evolved in time considering benefit for customers, flexibility, and expanding capabilities.  I have tried to update the PR main comment at the very top to reflect the latest decisions on our discussions.  The internal Tasks and User Story that is tracking all these changes and rescoping I cannot change because AFAIK they are classified as Confidential Microsoft information.  However, I'll keep updating the PR's main comment at the top to reflect the latest proposals.

ms-RistoRK replied:
I have updated the code to use an event-based approach and disable dynamic components iff there is no hovering nor proximity occurring.  Please re-review PR.

AMollis reviewed (changes requested):
Thank you for the changes @ms-RistoRK !   

I still have some suggestions, and would like the opinions of the other Maintainers (@keveleigh , @shaynie , @david-c-kline , @whebertML , @ghazen-ml )

ms-RistoRK commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -174,6 +183,16 @@ public SpaceMode DistanceSpaceMode
+        /// <summary>
+        /// Indicates the number of duples Collider + Interactor that have been registered as triggering proximity.
+        /// </summary>
+        public int ActiveColliderWithInteractorCount => activeCollidersWithInteractor.Count;
```
Removed in latest commit.

ms-RistoRK commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -623,6 +654,59 @@ private float IntersectQuad(Vector3 start, Vector3 end, float depth)
+        /// <param name="xrBaseInteractor">Interactor that in combination with the collider was triggering proximity.</param>
+        public void UnregisterActiveColliderWithInteractor(Collider collider, XRBaseInteractor xrBaseInteractor)
+        {
+            if (collider != null && activeCollidersWithInteractor.Contains((collider, xrBaseInteractor)))
```
I forgot that behavior of the HashSets, thank you for catching that and letting me know.  Updated in latest commit.

ms-RistoRK commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -623,6 +654,59 @@ private float IntersectQuad(Vector3 start, Vector3 end, float depth)
+        /// <param name="xrBaseInteractor">Interactor triggering proximity.</param>
+        public void RegisterActiveColliderWithInteractor(Collider collider, XRBaseInteractor xrBaseInteractor)
+        {
+            if (collider != null && !activeCollidersWithInteractor.Contains((collider, xrBaseInteractor)))
```
I forgot that behavior of the HashSets, thank you for catching that and letting me know.  Updated in latest commit.
I implemented:
```cs
if (collider != null &&
    activeCollidersWithInteractor.Add((collider, xrBaseInteractor)) &&
    activeCollidersWithInteractor.Count >= 1)
{
   SetEnabledDynamicComponents(true);
}
```
instead of:
`activeCollidersWithInteractor.Count == 1`

because there may be more than one type of combinations collider+xrBaseInteractor that are triggering proximity.
Updated in latest commit.

ms-RistoRK commented on org.mixedrealitytoolkit.input/InteractionModes/INearInteractionModeTarget.cs:
```diff
@@ -0,0 +1,32 @@
+        /// </summary>
+        /// <param name="collider">Collider triggering proximity.</param>
+        /// <param name="xrBaseInteractor">Interactor triggering proximity.</param>
+        public void RegisterActiveColliderWithInteractor(Collider collider, XRBaseInteractor xrBaseInteractor);
```
Updated in latest commit.  Note: I actually used `ProximityEnteredEventArgs` and `ProximityExitedEventArgs` instead of `ProximityEnteredArgs` and `ProximityExitedArgs` because the first names are more similar to the existing code.  Resolving this conversation.

ms-RistoRK commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -623,6 +654,59 @@ private float IntersectQuad(Vector3 start, Vector3 end, float depth)
+        /// Sets enable (true) or disable (false) the components that are in the ProximityEnabledComponents array.
+        /// </summary>
+        /// <param name="enable">True to enable the components and false otherwise.</param>
+        public void SetEnabledDynamicComponents(bool enable)
```
Implemented in latest commit.  Please re-review.

AMollis commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -174,11 +197,26 @@ public SpaceMode DistanceSpaceMode
+        /// <summary>
+        /// Indicates whether anything is hovering on the button (true) or not (false).
+        /// </summary>
+        private bool isHovered = false;
```
This is hiding the base isHovered variable.

You should be able to just use the base IsHovered value.

ms-RistoRK replied:
I missed that, thanks for catching it and letting me know.  I removed it in the latest commit.  Resolving this.

AMollis commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -462,7 +510,10 @@ protected override void OnHoverEntered(HoverEnterEventArgs args)
                 }
 
                 validPokeInteractors.Add(pokeInteractor);
+                isHovered = true;
```
We should be reusing the base `IsHovered` value

ms-RistoRK replied:
I missed that, thanks for catching it and letting me know.  I removed it in the latest commit.  Resolving this.

AMollis commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -623,6 +681,85 @@ private float IntersectQuad(Vector3 start, Vector3 end, float depth)
+                activeCollidersWithInteractor.Count >= 1)
+            {
+                isInProximity = true;
+                ProximityHoverEntered?.Invoke(this, proximityEnteredEventArgs, null);
```
See my comment with the `UpdateProximityHovered()` suggestion....you should remove the event invoking here and call `UpdateProximityHovered()`

ms-RistoRK replied:
Implemented in latest commit.  Resolving this.

AMollis commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -142,12 +147,30 @@ public SpaceMode DistanceSpaceMode
+        /// <summary>
+        /// Fired when an interactable starts hovering or becomes in proximity of the button.
+        /// </summary>
+        [field: SerializeField, Tooltip("Fired when an interactable starts hovering or becomes in proximity of the button.")]
```
We should avoid having multiple arguments in an event.    Please take a look at the hover events for the pattern we should use.   Look at the firstHoverEntered and lastHoverExited events (https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@2.0/api/UnityEngine.XR.Interaction.Toolkit.XRBaseInteractable.html).   Your event should have a single argument. For example, the HoverEnterEvent as a single HoverEnterEventArgs (https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@2.0/api/UnityEngine.XR.Interaction.Toolkit.HoverEnterEvent.html).

Also, for this scenario, we're more interested in the first and last proximity hovers ...  since we are only wanting to hide the front plates.  
 
So here's what I suggest.  Add two event types and args types

```cs
/// <summary>
/// The event that is called only when the first Interactor begins a proximity hover
/// over this Interactable. Subsequent Interactors that begin a proximity over this Interactable
/// will not cause this event to be invoked as long as any others are still proximity hovering.
/// </summary>
[Serializable]
public sealed class ProximityHoverEnterEvent : UnityEvent<ProximityHoverEnterEventArgs>

/// <summary>
/// Event data associated with the event when an Interactor initiates a proximity hover over an Interactable.
/// </summary>
public class ProximityHoverEnterEventArgs 
{
    /// <summary>
    /// The Interactable associated with the interaction event.
    /// </summary>
    public new INearInteractionModeTarget interactableObject
    {
        get => (IXRHoverInteractable)base.interactableObject;
        set => base.interactableObject = value;
    }
 }

/// <summary>
/// The event that is called only when the last remaining proximity hovering Interactor
/// ends proximity hovering over this Interactable.
/// </summary>
[Serializable]
public sealed class ProximityHoverExitEvent : UnityEvent<ProximityHoverExitEventArgs>

/// <summary>
/// Event data associated with the event when an Interactor initiates a proximity hover over an Interactable.
/// </summary>
public class ProximityHoverExitEventArgs 
{
    /// <summary>
    /// The Interactable associated with the interaction event.
    /// </summary>
    public new INearInteractionModeTarget interactableObject
    {
        get => (IXRHoverInteractable)base.interactableObject;
        set => base.interactableObject = value;
    }
 }
```       
Then add two events like this

```cs
/// <summary>
/// The event that is called only when the first Interactor begins a proximity hover
/// over this Interactable. Subsequent Interactors that begin a proximity over this Interactable
/// will not cause this event to be invoked as long as any others are still proximity hovering.
/// </summary>
[field: SerializeField, Tooltip("The event that is called only when the first Interactor begins a proximity hover over this Interactable..")]
public ProximityHoverEnterEvent FirstProximityHoverEntered { get; set } = new ProximityHoverEnterEvent();

/// <summary>
/// The event that is called only when the last remaining proximity hovering Interactor
/// ends proximity hovering over this Interactable.
/// </summary>
[field: SerializeField, Tooltip("The event that is called only when the last remaining proximity hovering Interactor ends proximity hovering over this Interactable.")]
public ProximityHoverEnterEvent LastProximityHoverEntered { get; set } = new ProximityHoverExitEvent();
```

Then property for querying proximity hover state.  Like this?

```cs
public bool IsProximityHovered { get; private set; } = false;
```

Then add a function to fire events as needed. Maybe like this?

```cs
private void UpdateProximityHovered()
{
   bool oldValue = IsProximityHovered;
   IsProximityHovered = IsHovered || activeCollidersWithInteractor.Count > 0;

   if (oldValue != IsProximityHovered)
   {
        if (IsProximityHovered) 
        {
              // Fire FirstProximityHoverEntered 
        }
        else
        {
            // Fire LastProximityHoverEntered
        }       
   }
}
```

ms-RistoRK replied:
Thank you so much for the detailed feedback.  I have attended your suggestions.  I adjusted the names a little bit to reflect that the events respond to a combined Proximity + Hovering state.

AMollis commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -623,6 +681,85 @@ private float IntersectQuad(Vector3 start, Vector3 end, float depth)
+                activeCollidersWithInteractor.Count == 0)
+            {
+                isInProximity = false;
+                ProximityHoverExited?.Invoke(this, proximityExitedEventArgs, null);
```
See my comment with the `UpdateProximityHovered()` suggestion....you should remove the event invoking here and call `UpdateProximityHovered()`

ms-RistoRK replied:
Implemented in latest commit.  Resolving this.

AMollis commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -623,6 +681,85 @@ private float IntersectQuad(Vector3 start, Vector3 end, float depth)
+        /// <param name="sender">The source of the event.</param>
+        /// <param name="proximityEnteredEventArgs">ProximityEnteredEventArgs if the event was triggered by proximity, null otherwise.</param>
+        /// <param name="hoverEnterEventArgs">HoverEnterEventArgs if the event was triggered by hovering, null otherwise.</param>
+        private void OnProximityHoverEntered(object sender, ProximityEnteredEventArgs proximityEnteredEventArgs, HoverEnterEventArgs hoverEnterEventArgs)
```
I don't think `OnProximityHoverEntered` `OnProximityHoverExited` and `SetEnabledDynamicComponents` are now required since we decided to use events only.

ms-RistoRK replied:
Removed when removed ProximityHoveringDynamicComponents array.

AMollis commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -474,7 +525,14 @@ protected override void OnHoverExited(HoverExitEventArgs args)
+                }
             }
+
+            ProximityHoverExited?.Invoke(this, null, args);
```
See my comment with the `UpdateProximityHovered()` suggestion....you should remove the event invoking here and call `UpdateProximityHovered()`

ms-RistoRK replied:
Implemented in latest commit.  Resolving this.

AMollis commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -474,7 +525,14 @@ protected override void OnHoverExited(HoverExitEventArgs args)
+
+                if (validPokeInteractors.Count == 0)
+                {
+                    isHovered = false;
```
We should be reusing the base `IsHovered` value

ms-RistoRK replied:
I missed that, thanks for catching it and letting me know.  I removed it in the latest commit.  Resolving this.

AMollis commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -462,7 +510,10 @@ protected override void OnHoverEntered(HoverEnterEventArgs args)
+                isHovered = true;
             }
+
+            ProximityHoverEntered?.Invoke(this, null, args);
```
See my comment with the `UpdateProximityHovered()` suggestion....you should remove the event invoking here and call `UpdateProximityHovered()`

ms-RistoRK replied:
Implemented in latest commit.  Resolving this.

AMollis commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -142,12 +147,30 @@ public SpaceMode DistanceSpaceMode
+        /// Array with Components (set in Editor) that will be enabled/disabled based on proximity and hovering.
+        /// </summary>
+        [field: SerializeField, Tooltip("Array with Components (set in Editor) that will be enabled/disabled based on proximity and hovering.")]
+        public Component[] ProximityHoveringDynamicComponents { get; set; } = null;
```
I was under the impression that we decided to remove this property given @whebertML 's suggestion as well.  

Let's start with adding events only and add property later if needed.

ms-RistoRK replied:
Aw :( ... I liked the idea of having the Component array because, imho, it simplifies making dynamic UIs with a tendency to improved rendering performance.  Such array empowers developers to quickly modify their scene and render UI components with rich visuals only when the user is about to interact with the associated UI elements and make a simple rendering when the user is not interacting with that part of the UI releasing GPU time for other tasks.  I was thinking on something like simplified UI rendering when the UI element is in the eye's Peripheral View and a rich rendering when the UI element is in the eye's Field of View.  Furthermore, such array, available in Editor would allow customers to quickly prototype and test their dynamic UIs for their specific needs.
On the other side, I do recognize you have more experience with MRTK3 framework and appreciate all you feedback and time you have invested in this PR, so I'll follow your lead; I'll remove the array next.

ms-RistoRK replied:
Removed in latest commit.  Resolving this.

AMollis commented on org.mixedrealitytoolkit.input/InteractionModes/INearInteractionModeTarget.cs:
```diff
@@ -0,0 +1,32 @@
+    /// <remarks>
+    /// This interface is needed to prevent a circular reference between MRTK Input and MRTK UX Core Scripts packages.
+    /// </remarks>
+    public interface INearInteractionModeTarget
```
(sorry of the not suggesting this sooner)....

However, we should rename this to `IProximityHoverInteractable` to match the patterns of XRI (e.g. `IXRHoverInteractable`

ms-RistoRK replied:
Got it, renamed in latest commit.

AMollis reviewed (changes requested):
It's looking good @ms-RistoRK ....

some more changes requested. Changes are mostly around following XRI patterns

AMollis reviewed (changes requested):
Code looks great! 

Can you please polish the PressableButton inspector window, and move the new events under the "MRTK Events" header?

![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/36461279/7260c9e7-7856-41a1-ab07-f0baf7dca07b)

Also, please make sure the that "UXComponent" package's dependency for the "UXCore" package is updated to the latest version....since you are taking on additional dependencies.  You need to updated for 3.1.0 to 3.2.0

![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/36461279/b3444daf-c9db-4d77-84d0-a211f5b07141)

ms-RistoRK commented:
> Code looks great!
> 
> Can you please polish the PressableButton inspector window, and move the new events under the "MRTK Events" header?
> 

Done in latest commit, Proximity-hover events are now under "MRTK Events" > "Proximity-Hover Events", for example:
![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/83b8b0e5-e4ce-4600-abe0-a5a28ae78b92)

AMollis reviewed (changes requested):
I apologize for this... :(

After reviewing the inspector changes, I just discovered the `TimedFlag` in MRTK... https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/blob/55e2f38f086b517fcfd562e6ab4d51062211079c/org.mixedrealitytoolkit.core/Utilities/TimedFlag.cs#L20C18-L20C27

The `TimedFlag` class has an Entered and Exited event.   We should probably follow the MRTK pattern here.   Please collapse `LastProximityHoverEntered/Exitted` and `public bool IsProximityHovered` into `public TimedFlag IsProximityHovered` . Sorry that I didn't know about this earlier.

Using the `TimedFlag` also means we no longer need the `XRProximityHoverEvents.cs`.  Again sorry about the late breaking discovery.

ms-RistoRK commented on org.mixedrealitytoolkit.uxcore/Button/PressableButton.cs:
```diff
@@ -142,12 +147,26 @@ public SpaceMode DistanceSpaceMode
+        /// ends proximity-hovering over this Interactable.
+        /// </summary>
+        [field: SerializeField, Tooltip("The event that is called only when the last remaining proximity-hovering Interactor ends proximity-hovering over this Interactable.")]
+        public ProximityHoverExitEvent LastProximityHoverEntered { get; set; } = new ProximityHoverExitEvent();
```
Switched to TimedFlag in latest commit, resolving this.

ms-RistoRK commented on org.mixedrealitytoolkit.core/Editor/Editors/BaseInteractableEditor.cs:
```diff
@@ -27,6 +27,9 @@ public class BaseInteractableEditor : XRBaseInteractableEditor
         private SerializedProperty disabledInteractorTypes;
 
+        private SerializedProperty FirstProximityHoverEntered;
+        private SerializedProperty LastProximityHoverEntered;
```
Switched to TimedFlag in latest commit, resolving this.

ms-RistoRK commented:
> I apologize for this... :(
> 
> After reviewing the inspector changes, I just discovered the `TimedFlag` in MRTK... https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/blob/55e2f38f086b517fcfd562e6ab4d51062211079c/org.mixedrealitytoolkit.core/Utilities/TimedFlag.cs#L20C18-L20C27
> 
> The `TimedFlag` class has an Entered and Exited event. We should probably follow the MRTK pattern here. Please collapse `LastProximityHoverEntered/Exitted` and `public bool IsProximityHovered` into `public TimedFlag IsProximityHovered` . Sorry that I didn't know about this earlier.
> 
> Using the `TimedFlag` also means we no longer need the `XRProximityHoverEvents.cs`. Again sorry about the late breaking discovery.

No worries, I wasn't aware that existed either :) ... and better find it now and change it now than having to change it once it is in Prod and consumed by our customers.  I made the change in the [latest commit](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/611/commits/2a1d3c2e46d17b58072bff31f37d74b23dafa6ed).  I also updated Editor's Inspector and the TimedFlag property + events now appear under StatefulInteractable Events as shown next:

![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/00f17947-9620-4e04-a5cb-a2e49fa5adcc)

AMollis commented on org.mixedrealitytoolkit.core/Interactables/StatefulInteractable.cs:
```diff
@@ -177,6 +177,12 @@ public string SpeechRecognitionKeyword
+        /// Is this object proximity-hovered by an active Collider + Interactor combination?
+        /// </summary>
+        [field: SerializeField, EditableTimedFlag, Tooltip("Is this object proximity-hovered by an active Collider + Interactor combination?")]
+        public TimedFlag IsProximityHovered { get; private set; } = new TimedFlag();
```
Let's move IsProximityHovered in the PressableButton.

I'm not sure it makes sense to be in StatefuleIteractable, since nothing in StatefuleIteractable will change this value.

ms-RistoRK replied:
That was my original approach too but since TimedFlag resides in MixedReality.Toolkit.Core binaries and its Active property is of type 'internal' I cannot change it if I declare it in PressableButton because that class exists in MixedReality.Toolkit.UXCore binaries and don't have access to 'internal' properties of other binaries.

ms-RistoRK replied:
Moved IsProximityHover property to PressableButton class and MixedReality.Toolkit.Core internals visible to MixedReality.Toolkit.UX assembly in [latest commit](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/611/commits/7a763b5a32c2b87e5f95aba1c4c147eead377387).

AMollis commented on org.mixedrealitytoolkit.uxcore/Editor/Inspectors/PressableButton/PressableButtonInspector.cs:
```diff
@@ -210,6 +213,23 @@ private float DrawPlaneAndHandle(Vector3[] vertices, Vector2 halfExtents, float
         static bool advancedButtonFoldout = false;
         static bool editorFoldout = false;
 
+        protected override void DrawMRTKInteractableFlags()
```
nit.   Add 

```cs
/// <inheritdoc />
```

To the top of the function, so the doc generation picks up the base description

ms-RistoRK replied:
Done in latest commit.

AMollis commented on org.mixedrealitytoolkit.core/Interactables/StatefulInteractable.cs:
```diff
@@ -320,7 +320,6 @@ public void ForceSetToggled(bool active)
             IsToggled.Active = active;
         }
 
```
Undo space changes to this file

ms-RistoRK replied:
Restored empty line in latest commit.

AMollis reviewed (approved):
Looks good.  I sign-off with some minor requests

AMollis reviewed (changes requested):
UXCore's dependencies need to be updated, since they now depend on a change to Core

Change UXCore's `package.json` dependency from:

```
  "org.mixedrealitytoolkit.core": "3.1.0",
```

to

```
  "org.mixedrealitytoolkit.core": "3.2.0",
```

ms-RistoRK commented:
> UXCore's dependencies need to be updated, since they now depend on a change to Core
> 
> Change UXCore's `package.json` dependency from:
> 
> ```
>   "org.mixedrealitytoolkit.core": "3.1.0",
> ```
> 
> to
> 
> ```
>   "org.mixedrealitytoolkit.core": "3.2.0",
> ```

Done in latest commit.

keveleigh commented on org.mixedrealitytoolkit.input/InteractionModes/BaseProximityEventArgs.cs:
```diff
@@ -0,0 +1,40 @@
+using System;
```
Missing copyright header

ms-RistoRK replied:
Added in https://github.com/ms-RistoRK/MixedRealityToolkit-Unity/commit/1b5f80d9957ebd2762a1f6752d656f656674b671 resolving this.

keveleigh commented on org.mixedrealitytoolkit.input/InteractionModes/ProximityEnteredEventArgs.cs:
```diff
@@ -0,0 +1,29 @@
+using UnityEngine;
```
Missing copyright header

ms-RistoRK replied:
Added in https://github.com/ms-RistoRK/MixedRealityToolkit-Unity/commit/1b5f80d9957ebd2762a1f6752d656f656674b671 resolving this.

keveleigh reviewed (commented):
I know I'm way late to this discussion, but I have some concerns about this PR

keveleigh commented on org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs:
```diff
@@ -20,10 +20,52 @@ public class NearInteractionModeDetector : ProximityDetector
+                        IXRProximityInteractable nearInteractionMode = collider.GetComponent<IXRProximityInteractable>();
+                        if (nearInteractionMode != null)
+                        {
+                            foreach (XRBaseInteractor xrBaseInteractor in nearInteractors)
```
Instead of sending multiple events to account for all `nearInteractors`, should we just send _one_ relaying info about either:

1. The `ProximityDetector` itself. We could potentially expose the list of `nearInteractors` as an `IReadOnlyList` or something if we think the interactables need that info
2. Specifically the `nearInteractor`(s) doing the selecting. They likely aren't all selecting at once, but one of the cases for this logic is that _any_ `nearInteractor` is selecting (`IsNearInteractorSelecting()`)

ms-RistoRK commented:
@keveleigh Thank you for the additional feedback.  I've created this [issue](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/issues/651) to track the implementation of your feedback.

ms-RistoRK commented on org.mixedrealitytoolkit.input/InteractionModes/BaseProximityEventArgs.cs:
```diff
@@ -0,0 +1,40 @@
+    /// <summary>
+    /// Event data associated with proximity events triggered by a Collider and an Interactor combo.
+    /// </summary>
+    public abstract partial class BaseProximityEventArgs : EventArgs
```
Removed in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/commit/01ff8cbd0ed639b74489cfb4eb13d06a7c163739

ms-RistoRK commented on org.mixedrealitytoolkit.input/InteractionModes/BaseProximityEventArgs.cs:
```diff
@@ -0,0 +1,40 @@
+        /// <summary>
+        /// The object that triggered the proximity event.
+        /// </summary>
+        public object sender { get; private set; }
```
Fixed in https://github.com/ms-RistoRK/MixedRealityToolkit-Unity/commit/23d1e55861b9559c0f9721f0fa9dcbe0c5dee808

ms-RistoRK commented on org.mixedrealitytoolkit.input/InteractionModes/BaseProximityEventArgs.cs:
```diff
@@ -0,0 +1,40 @@
+        /// <param name="sender">Source of event.</param>
+        /// <param name="collider">Collider that triggers proximity event.</param>
+        /// <param name="interactor">XRBaseInteractor that triggers proximity event.</param>
+        public BaseProximityEventArgs(object sender, Collider collider, XRBaseInteractor interactor)
```
Not really in this case.  I am just used to do it as in most of C# apps I have seen that's the pattern: https://learn.microsoft.com/en-us/dotnet/standard/events/ ... removed in https://github.com/ms-RistoRK/MixedRealityToolkit-Unity/commit/21fc02f3128b67951de8a79618e7506c46cdb947

ms-RistoRK commented on org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs:
```diff
@@ -20,10 +20,52 @@ public class NearInteractionModeDetector : ProximityDetector
+                    Collider previouslyDetectedCollider = previouslyDetectedColliders[i];
+                    if (!DetectedColliders.Contains(previouslyDetectedCollider) && previouslyDetectedCollider != null)
+                    {
+                        IXRProximityInteractable nearInteractionMode = previouslyDetectedCollider.GetComponent<IXRProximityInteractable>();
+                        if (nearInteractionMode != null)
```
Thank you for the feedback and additional explanation.  I updated the code to use TryGetComponent instead of GetComponent in https://github.com/ms-RistoRK/MixedRealityToolkit-Unity/commit/8ba56f4008476f7a76a02d552321ec002b9ad181

ms-RistoRK commented on org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs:
```diff
@@ -20,10 +20,52 @@ public class NearInteractionModeDetector : ProximityDetector
+        /// colliders stopped being detected and update their buttons front plate and rounded
+        /// rect if they are labeled as dynamic (based on proximity).
+        /// </summary>
+        List<Collider> previouslyDetectedColliders = new List<Collider>();
```
Thank you for noticing and letting me know.  I added the missing access modifier in https://github.com/ms-RistoRK/MixedRealityToolkit-Unity/commit/b0ade90ada993946cf351d7cbcbe5fac55db3799

ms-RistoRK commented on org.mixedrealitytoolkit.input/InteractionModes/ProximityEnteredEventArgs.cs:
```diff
@@ -0,0 +1,29 @@
+            //Empty on purpose
+        }
+
+        /// <summary>
+        /// The collider associated with the proximity entered event.
+        /// </summary>
+        public new Collider collider => base.collider;
+
+        /// <summary>
+        /// The Interactor associated with the proximity entered event.
+        /// </summary>
+        public new XRBaseInteractor interactor => base.interactor;
```
Removed in https://github.com/ms-RistoRK/MixedRealityToolkit-Unity/commit/df73d25ac533f5492e4f2f57aa19c4e9a9264192

ms-RistoRK commented on org.mixedrealitytoolkit.input/InteractionModes/ProximityExitedEventArgs.cs:
```diff
@@ -0,0 +1,29 @@
+            //Empty on purpose
+        }
+
+        /// <summary>
+        /// The Collider associated with the proximity exited event.
+        /// </summary>
+        public new Collider collider => base.collider;
+
+        /// <summary>
+        /// The Interactor associated with the proximity exited event.
+        /// </summary>
+        public new XRBaseInteractor interactor => base.interactor;
```
Removed in https://github.com/ms-RistoRK/MixedRealityToolkit-Unity/commit/df73d25ac533f5492e4f2f57aa19c4e9a9264192

ms-RistoRK commented on org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs:
```diff
@@ -20,10 +20,52 @@ public class NearInteractionModeDetector : ProximityDetector
+                {
+                    if (!previouslyDetectedColliders.Contains(collider))
+                    {
+                        IXRProximityInteractable nearInteractionMode = collider.GetComponent<IXRProximityInteractable>();
+                        if (nearInteractionMode != null)
+                        {
+                            foreach (XRBaseInteractor xrBaseInteractor in nearInteractors)
+                            {
+                                collider.GetComponent<IXRProximityInteractable>().OnProximityEntered(new ProximityEnteredEventArgs(this, collider, xrBaseInteractor));
```
Flipped conditional and cycle, and reused query result in https://github.com/ms-RistoRK/MixedRealityToolkit-Unity/commit/f6b2ddfc4aa12e10b58c3bee2c68f52c32785fe7
